File: go/2.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func Win(a int) int {
	return a + 6
}
func Draw(a int) int {
	return a + 3
}

func getWin(a int) int {
	if a+1 < 4 {
		return a + 1
	} else {
		return 1
	}
}

func calc(a, b int) int {
	return (3 + a - b) % 3
}

func getLose(a int) int {
	if a-1 > 0 {
		return a - 1
	} else {
		return 3
	}
}

func comp1(game []int) int {
	if game[0] == game[1] {
		return Draw(game[1])
	}
	if calc(game[0], game[1]) == 1 {
		return game[1]
	} else {
		return Win(game[1])
	}
}

func comp2(game []int) int {
	if game[1] == 1 {
		return getLose(game[0])
	}
	if game[1] == 2 {
		return Draw(game[0])
	} else {
		return Win(getWin(game[0]))
	}
}

func sum(a, b int) int {
	return a + b
}

func format(guide string) []int {
	re := regexp.MustCompile("A|X").ReplaceAllString(guide, "1")
	re = regexp.MustCompile("B|Y").ReplaceAllString(re, "2")
	re = regexp.MustCompile("C|Z").ReplaceAllString(re, "3")
	s := strings.Split(re, " ")

	opp, err := strconv.Atoi(s[0])
	check(err)
	pl, er := strconv.Atoi(s[1])
	check(er)

	res := []int{opp, pl}

	return res
}

func read() []string {
	file, err := os.ReadFile("../src/2.txt")
	check(err)

	return strings.Split(string(file), "\n")
}

func main() {
	var formatted1 []int
	var formatted2 []int

	for _, val := range read() {
		formatted1 = append(formatted1, comp1(format(val)))
	}
	for _, val := range read() {
		formatted2 = append(formatted2, comp2(format(val)))
	}

	var sum1 int
	var sum2 int

	for _, val := range formatted1 {
		sum1 += val
	}
	for _, val := range formatted2 {
		sum2 += val
	}

	fmt.Println("Part 1:", sum1)
	fmt.Println("Part 2:", sum2)

}

File: py/2.py
import re
from functools import reduce


f = open("../src/2.txt")
file = f.read().splitlines()


def Win(a): return a+6
def Draw(a): return a+3
def Lose(a): return a


def getWin(a): return a+1 if a+1 < 4 else 1
def getLose(a): return a-1 if a-1 > 0 else 3


def calc(a, b): return (3+a-b) % 3


def comp1(game):
    if game[0] == game[1]:
        return Draw(game[1])
    if (calc(game[0], game[1]) == 1):
        return Lose(game[1])
    else:
        return Win(game[1])


def comp2(game):
    if game[1] == 1:
        return Lose(getLose(game[0]))
    if game[1] == 2:
        return Draw(game[0])
    else:
        return Win(getWin(game[0]))


def format(guide):
    res = re.sub(r"(A)|(X)", "1", guide)
    res = re.sub(r"(B)|(Y)", "2", res)
    res = re.sub(r"(C)|(Z)", "3", res)
    return [int(res[0]), int(res[2])]


def sum(a, b): return a+b


sum1 = list(map(comp1, list(map(format, file))))
sum2 = list(map(comp2, list(map(format, file))))

print("Part 1 Score:", reduce(sum, sum1, 0))
print("Part 2 Score:", reduce(sum, sum2, 0))
